snapcraft/linters/library_linter.py

```diff
@@ -98,21 +116,92 @@ def _check_dependencies_satisfied(
+                    text=f"missing dependency {dependency.name!r}.",
                     url="https://snapcraft.io/docs/linters-library",
                 )
                 issues.append(issue)
+
+    def _get_unused_library_issues(
+        self, all_libraries: Set[Path], used_libraries: Set[Path]
+    ) -> List[LinterIssue]:
+        """Get a list of unused library issues.
+
+        :param all_libraries: a set of paths to all libraries
+        :param used_libraries: a set of libraries used by elf files in the snap
+
+        :returns: list of LinterIssues for unused libraries
+        """
+        issues: List[LinterIssue] = []
+        unused_libraries = all_libraries - used_libraries
+
+        # sort libraries so the results are ordered in a determistic way
+        for library_path in sorted(unused_libraries):
+            try:
+                # Resolving symlinks to a library will change the path from relative
+                # to absolute. To make it relative again, remove the current directory
+                # prefix from the path.
+                resolved_library_path = library_path.resolve().relative_to(Path.cwd())
+            except ValueError:
+                # A ValueError is not expected because these libraries should be within
+                # the current directory, but check anyways
+                emit.debug(f"could not resolve path for library {library_path!r}")
+                continue
+
+            library = ElfFile(path=resolved_library_path)
+
+            # skip linting files listed in the ignore list
+            if self._is_file_ignored(library):
+                continue
+
+            issue = LinterIssue(
+                name=self._name,
+                result=LinterResult.WARNING,
+                filename=library.soname,
+                text=f"unused library {str(library.path)!r}.",
+                url="https://snapcraft.io/docs/linters-library",
+            )
+            issues.append(issue)
+
+        return issues
+
+    def _is_library_path(self, path: Path) -> bool:
+        """Check if a file is in a library directory.
+
+        An elf file is considered a library if it has an soname and it is within a
+        library directory.
+        A library directory is a directory whose pathname ends in `lib/` or
+        `lib/<architecture-triplet>/` (i.e. lib/x86_64-linux-gnu/`)
+
+        :param path: filepath to check
+
+        :returns: True if the file is in a library directory. False if the path is not
+        a file or if it is not in a library directory.
+        """
+        # follow symlinks
+        path = path.resolve()
+
+        if not path.is_file():
+            return False
+
+        # TODO: get library directories from LD_LIBRARY_PATH and `ld --verbose`
+        # instead of matching against patterns
+        if path.match("lib/*") or path.match("lib32/*") or path.match("lib64/*"):
+            return True
+
+        for architecture in elf_utils.ARCH_CONFIG.values():
```

Comment:
Ideally we should keep `ARCH_CONFIG` internal to elf_utils, maybe we could add a function that returns a list of valid triplets to use here?

Reply:
Made a function to return valid triplets and wrote unit tests.